crates/node/src/lib.rs

```diff
@@ -287,4 +276,81 @@ impl Node {
 
         Ok(())
     }
+
+    async fn upsert_node_config(
+        metadata_store_client: MetadataStoreClient,
+        options: &Options,
+    ) -> Result<NodesConfiguration, ReadModifyWriteError> {
+        // todo: Make upsert timeout configurable
+        let retry_strategy = ExponentialBackoff::from_millis(10)
+            .max_delay(Duration::from_secs(5))
+            .factor(2)
+            .map(jitter)
+            .take(15);
+        let upsert_start = Instant::now();
+
+        RetryIf::spawn(
+            retry_strategy,
+            || {
+                metadata_store_client.read_modify_write(NODES_CONFIG_KEY.clone(), |nodes_config| {
+                    let mut nodes_config = nodes_config.unwrap_or_else(|| {
+                        NodesConfiguration::new(Version::INVALID, options.cluster_name.clone())
+                    });
+
+                    // check whether we have registered before
+                    let node_config = nodes_config.find_node_by_name(&options.node_name).cloned();
+
+                    let my_node_config = if let Some(mut node_config) = node_config {
+                        assert_eq!(options.node_name, node_config.name, "node name must match");
+
+                        // update node_config
+                        node_config.roles = options.roles;
+                        node_config.address = options.server.advertise_address.clone();
+                        node_config.current_generation.bump_generation();
```

Comment:
Ideally, we should avoid ping-pong pre-emption if we saw that someone else has incremented the generation for our node id between retries. That'd be possible if this closure keeps state between retries.

Reply:
True. I guess in this situation, the node should terminate itself because it detected another node with the same id to be active.